### PR TITLE
oshmem: fix race condition on new contexts

### PR DESCRIPTION
**Pull request by janjust**
Do not add private contexts to active list, they need only be visible to user

Signed-off-by: Tomislav Janjusic <tomislavj@mellanox.com>
Signed-off-by: Artem Y. Polyakov <artemp@mellanox.com>

### CODE REVIEW

**Comment by janjust** [comment]
@jladd-mlnx fyi

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 1]
```diff
@@ -698,6 +698,11 @@ int mca_spml_ucx_ctx_create(long options, shmem_ctx_t *ctx)
         opal_progress_register(spml_ucx_ctx_progress);
     }
 
+    if (options & SHMEM_CTX_PRIVATE) {
```
Comment (brminich) [inline oshmem/mca/spml/ucx/spml_ucx.c]: minor: could avoid code duplication if handle ` if (!(options & SHMEM_CTX_PRIVATE)) ` instead

**Comment by hppritcha** [comment]
bot:ompi:retest

**Comment by artpol84** [comment]
@brminich the race is when the private context that is not lock-protected is being progressed bo some other thread as part of opal progress while the thread that owns this private context is accessing it.

File: oshmem/mca/spml/ucx/spml_ucx.c (hunk 1 above)
Comment (artpol84) [reply oshmem/mca/spml/ucx/spml_ucx.c]: Good point, @janjust you can put 
https://github.com/open-mpi/ompi/pull/7065/files#diff-8af288f2560951b1798fcbf1131caa78L701-L703
into the if statement as 
https://github.com/open-mpi/ompi/pull/7065/files#diff-8af288f2560951b1798fcbf1131caa78R702-R703
and https://github.com/open-mpi/ompi/pull/7065/files#diff-8af288f2560951b1798fcbf1131caa78R710-R711
are essentially the same.

**Comment by artpol84** [comment]
Per discussion with @manjugv:
A private context should only be progressed by the thread that has created it. This allows avoiding locking around it to get the best performance.
This PR excludes private contexts from being included in the list of contexts to be processed during opal progress so it won't be implicitly progressed by an arbitrary thread calling progress.

**Comment by janjust** [comment]
PR isn't ready yet, don't merge

**Comment by artpol84** [comment]
@brminich please have a look.
This PR addresses 2 issues now:
* Race condition for PRIV contexts
* Memory leak when a large number of contexts is created/destroyed. Now we reuse released contexts when possible.

**Comment by janjust** [comment]
@artpol84 pushed the latest, and tested

**Comment by hppritcha** [comment]
@janjust is this PR ready to merge now?

**Comment by janjust** [comment]
@hppritcha yes

**Pull request merged**